src/v/archival/ntp_archiver_service.cc

```diff
@@ -2936,6 +2936,34 @@ ss::future<bool> ntp_archiver::do_upload_local(
       .metadata_size_hint = tx_size,
     };
 
+    const bool checks_disabled
+      = config::shard_local_cfg()
+          .cloud_storage_disable_upload_consistency_checks.value();
+
+    if (!checks_disabled && upl_res.has_record_stats()) {
+        auto stats = upl_res.record_stats();
+        // Validate segment content. The 'stats' is computed when
+        // the actual segment is scanned and represents the 'ground truth' about
+        // its content. The 'meta' is the expected segment metadata. We
+        // shouldn't replicate it if it doesn't match the 'stats'.
+        if (
+          meta.size_bytes != stats.size_bytes
+          || meta.base_offset != stats.base_rp_offset
+          || meta.committed_offset != stats.last_rp_offset) {
+            vlog(
```

Comment:
nit: this seems to be copied directly from `wait_uploads()` call. These seem like universal checks we'd want for any segment. Perhaps pull this out into a method and reuse it?

Reply:
done